lib/ddtrace/contrib/racecar/patcher.rb

```diff
@@ -33,28 +36,23 @@ def patched?
             @patched = false
           end
 
-          def start(event, _, payload)
-            ensure_clean_context!
+          def unpatch
```

Comment:
I see this being referenced in the tests. What is it used for?

Reply:
`patch` modifies global state, and only does anything the first time around – so the setup code we need to e.g. pass the `tracer` to the Subscriber instances won't be called on the second go.

Reply:
Gotcha. So this is particularly for testing purposes?

Reply:
Sure. How is this handled in other specs? I'd assume that, since `patch` is a no-op on the second call, this would be a general problem?

Reply:
In the general sense, it's usually not an issue because most integrations only monkey patch code, send some metadata, so we don't need to redo it each time for each test. There's something to be said about not carrying over any state between test runs, which is normally really important/preferable. However, it's not really feasible to truly un-monkey patch code and reset back to actual original state given the nature of this meta-programming.

Although in the case of Racecar, the patching is also subscribing to some instrumentation, we shouldn't really need to redo `patch` per test, no? Since the subscriptions should still be good across example runs. (I believe the specs were passing before we had the `unpatch`.)

Reply:
The problem was that the `tracer` instance is per-test, but once patched, the first tracer is permanently baked into the patch.  I don't know why this hasn't been a problem until now.

Reply:
Ahhh, okay I see. Because when we patch, we create a Subscriber with the reference to the tracer, and subscribe it to AS::N. Then that Subscriber object remains subscribed to AS::N with the reference to the old tracer, even as a new example starts running.

Perhaps we should make the Subscriber objects accessible, such that we can modify/unsubscribe them in the tests. Maybe something like:

**contrib/active_support/notifications.rb**

```ruby
self.subscribe(pattern, subscriber)
  ::ActiveSupport::Notifications.subscribe(pattern, subscriber)
end

self.unsubscribe(subscriber)
  ::ActiveSupport::Notifications.unsubscribe(subscriber)
end

self.subscribe!(pattern, span_name, options = {}, tracer = Datadog.tracer, &block)
  subscriber(tracer, span_name, options, &block).tap do |subscriber|
    subscribe(pattern, subscriber)
  end
end

self.subscriber(span_name, options = {}, tracer = Datadog.tracer, &block)
  Subscriber.new(tracer, span_name, options, &block)
end
```

**contrib/racecar/patcher.rb**

```ruby
def patch
  #...
  subscribe!
end

def subscribe!
  subscribers.each do |pattern, subscriber|
    Contrib::ActiveSupport::Notifications.subscribe(pattern, subscriber)
  end
end

def subscribers
  @subscribers ||= begin
    tracer = configuration[:tracer]
    {
      'process_message.racecar' => Contrib::ActiveSupport::Notifications.subscriber(NAME_MESSAGE, {}, tracer, &method(:process))
      'process_batch.racecar' => Contrib::ActiveSupport::Notifications.subscriber(BATCH_MESSAGE, {}, tracer, &method(:process))
    }
  end
end
```

**racecar_spec.rb**

```ruby
before(:each) do
  Contrib::Racecar::Patcher.subscribers.each do |_pattern, subscriber|
    subscriber.instance_variable_set(:@tracer, tracer)
  end
end
```

Although I think the above should work, I also think this case highlights a weakness in the current Racecar implementation; that once configured, you cannot reconfigure Racecar with a different tracer.

We could go a step further, and try to fix this by re-subscribing to AS::N with new Subscribers everytime the configuration changes, or maybe by doing some kind of lazy evaluation of configurable options like `tracer` (both of which are kind of complicated.) Something to think about maybe?

Reply:
I think it's actually possible to keep adding subscribers, it's just that `patch` short-circuits before getting to that part:

https://github.com/dasch/dd-trace-rb/blob/ce6b30cf66435fa88f7ee191549a99549c4324f5/lib/ddtrace/contrib/racecar/patcher.rb#L17-L23

I think a better solution would be to manage the patch status in a separate component and just let `patch` do its thing regardless of whether it's already been applied. It seems that the idempotency mechanics are the same everywhere, i.e. a class instance variable `@patched` that gets set to `true`.

Reply:
If we keep adding subscribers, won't we collect a bunch of old subscribers that errantly writing redundant traces? I'm worried about potential side effects from that.

If that's not much of a concern for these tests, then I perhaps I might suggest a simpler, lighter approach? By just undoing the `@patched` manually in the tests?

```
before(:each) do
  # Do this before calling Datadog.configure
  Contrib::Racecar::Patcher.instance_variable_set(:@patched, false)
end
```